stdlib/domain.ml

```diff
@@ -172,17 +92,33 @@ let do_before_first_spawn () =
     first_spawn_function := (fun () -> ())
   end
 
-let at_exit_key = DLS.new_key (fun () -> (fun () -> ()))
+module IMap = Map.Make (struct type t = int let compare = (-) end)
+let at_exit_callbacks = ref IMap.empty
+let at_exit_mutex = Mutex.create () (* protects [at_exit_callbacks] *)
 
 let at_exit f =
-  let old_exit : unit -> unit = DLS.get at_exit_key in
-  let new_exit () =
-    f (); old_exit ()
-  in
-  DLS.set at_exit_key new_exit
+  let domain_id = (self () :> int) in
+  Mutex.protect at_exit_mutex (fun () ->
+    let m = !at_exit_callbacks in
+    let old_exit_opt = IMap.find_opt domain_id m in
```

Comment:
I find this implementation surprisingly complex and I wonder if this is the right approach.

1. Maybe we actually want to have `Thread.at_exit`, which would be naturally easier to implement on top of `Thread_local_storage`? (Note: even if we did, we probably want to keep a `Domain.at_exit` around with the current semantics, but that would mostly be for advanced out-of-tree experiments.)

2. We discussed a way to implement DLS on top of TLS, isn't this function a good opportunity to try this? For example, it would seem fairly natural to me to implement this as a `(unit -> unit) ref Thread_local_storage.t` that is shared across all threads of a given domain. Am I missing something? (If we try, and it appears that DLS-on-top-of-TLS is worse than we thought, maybe that is a reason to reconsider.)

Reply:
I'd like to understand what is complex with this implementation. It is a single piece of shared state `at_exit_callbacks` protected by a mutex. 

If we want to retain `Domain.at_exit`, I'd rather keep the current implementation, and think about `Thread.at_exit` in a separate PR.

Reply:
Part of the argument we used to move DLS to TLS is that it "should be easy" to implement DLS on top of TLS.

But then, on the one time where we do need an implementation of DLS inside the compiler distribution, we punt and implement a globally-locked map of domain identifiers. Is this our recommendation for users as well, that they implement a globally-locked of domain identifiers? If not, why don't we follow our own recommendations?

I think that we should do our homework and implement `Domain.at_exit` as DLS on top of TLS. I see two possible outcomes:
1. If it turns out to be simple and pleasant, this would be better I think than your current implementation. Plus we have an example of DLS-over-TLS implementation we can point people at. Win!
2. If it turns out to be harder than we thought, or have performance downsides that we find problematic, then it suggests that our initial intuition was wrong and maybe we should reconsider or at least think harder. I would rather figure this out now than after a release.

Reply:
> Part of the argument we used to move DLS to TLS is that it "should be easy" to implement DLS on top of TLS.

The lazy-based implementation (https://github.com/ocaml/ocaml/pull/12719#issuecomment-1832179117) has a concurrency issue in that multiple threads on a domain may concurrently perform `DLS.get` which will concurrently force the lazy value, which is not concurrency safe. 

I haven't investigated the concurrent hash map + caching solution. This requires a concurrent hash map in the standard library @gadmm? or am I misunderstanding the solution? Regardless of the solution that we provide, there needs to be concurrency control on `DLS.set` (which must be preserved for backwards compatibility).